src/extensions/scratch3_wedo2/index.js

```diff
@@ -278,63 +350,223 @@ class WeDo2 {
-        this._socket.off('disconnect', this._onDisconnect);
+    _onMessage (base64) {
+        const data = Base64Util.base64ToUint8Array(base64);
+
+        if (data.length === 2) { // disconnect sensor
+            const connectID = data[0];
+            // zero out tilt
+            if (this._ports[connectID - 1] === WeDo2Types.TILT) {
+                this._sensors.tiltX = this._sensors.tiltY = 0;
+            }
+            // zero out distance
+            if (this._ports[connectID - 1] === WeDo2Types.DISTANCE) {
+                this._sensors.distance = 0;
+            }
+            // remove references to ports and motors
+            if (connectID === 1 || connectID === 2) {
+                this._ports[connectID - 1] = 'none';
+                this._motors[connectID - 1] = null;
+                // log.info(`this._ports = ${this._ports}`);
+                // log.info(`this._motors = ${this._mtors}`);
+            }
+        }
+
+        if (data.length === 3) { // distance sensor value?
+            this._sensors.distance = data[2];
+        }
+
+        if (data.length === 4) { // tilt sensor value?
+            this._sensors.tiltX = data[2];
+            this._sensors.tiltY = data[3];
+        }
+
+        if (data.length === 12) { // attached io?
+
+            const connectID = data[0];
+            const type = data[3];
+
+            // Record which port is connected to what type of device
+            if (connectID === 1 || connectID === 2) {
+                this._ports[connectID - 1] = type;
+            }
+
+            // Motor
+            if (type === WeDo2Types.MOTOR) {
+                this._motors[connectID - 1] = new WeDo2Motor(this, connectID - 1);
+            }
+
+            // Tilt Sensor
+            if (type === WeDo2Types.TILT) {
+                const cmd = new Uint8Array(11);
+                cmd[0] = 1; // sensor format
+                cmd[1] = 2; // command type: write
+                cmd[2] = connectID; // connect id
+                cmd[3] = WeDo2Types.TILT; // type
+                cmd[4] = 0; // mode: angle
+                cmd[5] = 1; // delta interval, 4 bytes
+                cmd[6] = 0;
+                cmd[7] = 0;
+                cmd[8] = 0;
+                cmd[9] = 1; // unit: pct?
+                cmd[10] = 1; // notifications enabled: true
+
+                this._send(UUID.INPUT_COMMAND, Base64Util.uint8ArrayToBase64(cmd))
+                    .then(() => {
+                        this._ble.read(UUID.IO_SERVICE, UUID.INPUT_VALUES, true, this._onMessage.bind(this));
+                    });
+            }
+
+            // Distance Sensor
+            if (type === WeDo2Types.DISTANCE) {
+                const cmd = new Uint8Array(11);
+                cmd[0] = 1; // sensor format
+                cmd[1] = 2; // command type: write
+                cmd[2] = connectID; // connect id
+                cmd[3] = WeDo2Types.DISTANCE; // type
+                cmd[4] = 0; // mode: detect
```

Comment:
I might want to put the modes up in an enum

Reply:
... but then I'd probably put the units and other things up there too and it'd be lots of enums, so I am still debating it